src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatter.cs

```diff
@@ -106,5 +111,19 @@ internal void Serialize(Stream serializationStream, object graph, bool check)
                 string assemblyName = FormatterServices.GetClrAssemblyName(t, out bool hasTypeForwardedFrom);
                 return new TypeInformation(FormatterServices.GetClrTypeFullName(t), assemblyName, hasTypeForwardedFrom);
             });
+
+        private static Exception CreateBinaryFormatterDisallowedException()
+        {
+            // If we're in an environment where BinaryFormatter never has a
+            // chance of succeeding, use PNSE. Otherwise use regular NSE.
+
+            string message = SR.BinaryFormatter_SerializationDisallowed;
+
+#if BINARYFORMATTER_PNSE
```

Comment:
Nit: The coding conventions prefer separate files over `#ifdefs`. This can actually look pretty good as separate file:

```
BinaryFormatter.PlatformNotSupported.cs

public sealed class BinaryFormatter : IFormatter
{
    public BinaryFormatter()
    {
    }

    public BinaryFormatter(ISurrogateSelector? selector, StreamingContext context)
    {
    }

    public object Deserialize(Stream serializationStream)
    {
        throw PNSE();
    }

    public void Serialize(Stream serializationStream, object graph)
    {
        throw PNSE();
    }
}
```

Reply:
Ended up using three files: one for the "real" Serialize / Deserialize methods, one for the PNSE Serialize / Deserialize methods, and one for the common logic like constructors and properties. Let me know if that's closer to what you were thinking.

Reply:
Note: because the PNSE files ripped out completely the real logic of the Serialize and Deserialize methods, this means that when compiling for Browser there are some internal fields across various types that are never touched and thus produce compiler warnings. I thought pragma suppressing all of them would be a bit too noisy, so instead I NoWarned that specific warning code _only when compiling against Browser_.

Reply:
> the common logic like constructors and properties

Ah ... I have not realized that there are also properties. Would it make sense to gut the whole type and throw from the constructor for the PNSE mode? (`s_typeNameCache` can be moved to `class TypeInformation`.)

I do not have a strong opinion about this.